crates/flowctl/src/source/unbundle.rs

```diff
@@ -0,0 +1,338 @@
+        }
+
+        (Ok(_), Existing::Overwrite) => {
+            tracing::info!(path = %flow_yaml_path.display(), "removing file because it already exists and `--existing=overwrite`");
+            fs::remove_file(flow_yaml_path)
+                .await
+                .context(format!("failed to remove: {}", flow_yaml_path.display()))?;
+            new_catalog
+        }
+        (Ok(_), Existing::MergeSpec) => {
+            tracing::info!(path = %flow_yaml_path.display(), "merging specs because it already exists and `--existing=merge-spec`");
+            let mut existing = parse_catalog_spec(flow_yaml_path).await.context(format!(
+                "reading existing catalog file: {}",
+                flow_yaml_path.display()
+            ))?;
+            merge_catalog(new_catalog, &mut existing)?;
+            existing
+        }
+        (Err(err), _) if err.kind() == std::io::ErrorKind::NotFound => new_catalog,
+        (Err(err), _) => {
+            anyhow::bail!(
+                "io error retrieving metadata for path: '{}': {}",
+                flow_yaml_path.display(),
+                err
+            );
+        }
+    };
+    Ok(Some(new_catalog))
+}
+
+struct Serializer {
+    json: bool,
+}
+
+impl Serializer {
+    pub fn to_vec<T: serde::Serialize>(&self, value: &T) -> anyhow::Result<Vec<u8>> {
+        // Our models embed serde_json's RawValue, which can only be serialized by serde_json.
+        // Do that first.
+        let v = serde_json::to_vec_pretty(value).context("serializing JSON")?;
+        if self.json {
+            Ok(v)
+        } else {
+            // We want YAML, so transcode directly from serialized JSON to YAML.
+            let mut v2 = Vec::new();
+            let mut deser = serde_json::Deserializer::from_slice(&v);
+            serde_transcode::transcode(&mut deser, &mut serde_yaml::Serializer::new(&mut v2))
+                .context("serializing YAML")?;
+
+            Ok(v2)
+        }
+    }
+}
+
+async fn try_write_file(path: &Path, contents: &[u8], existing: Existing) -> anyhow::Result<()> {
+    // Is the file allowed to exist before we create it?
+    let create_new = match existing {
+        Existing::Abort | Existing::Keep => true,
+        _ => false,
+    };
+    let result = fs::OpenOptions::new()
+        .create_new(create_new)
+        .create(true)
+        .write(true)
+        .open(path)
+        .await;
+    match result {
+        Ok(mut file) => {
+            file.write_all(contents).await?;
+            Ok(())
+        }
+        Err(err)
+            if err.kind() == std::io::ErrorKind::AlreadyExists && existing == Existing::Keep =>
+        {
+            tracing::info!(path = %path.display(), "skipping writing file because it already exists and --existing=skip");
+            Ok(())
+        }
+        // This branch also handles the case of `Existing::Error` if the file already exists.
+        Err(err) => Err(anyhow::anyhow!("unable to open file: {}", err)),
```

Comment:
Probably should include the file path here like you did above, just to avoid annoyingly hard to debug error messages. Or is the idea that `err` will contain the path?

Reply:
The path is added to the context in `write_file`. `try_write_file` was just created so that `write_file` only has to add that context in a single place instead of attaching it to both fallible calls.

Reply:
Ahh ok, missed that!